src/GZCTF/Controllers/GameController.cs

```diff
@@ -897,6 +897,24 @@ public async Task<IActionResult> Submit([FromRoute] int id, [FromRoute] int chal
         if (context.Result is not null)
             return context.Result;
 
+        var instance = await gameInstanceRepository.GetInstance(context.Participation!, challengeId, token);
+
+        if (instance is null)
+            return NotFound(new RequestResponse(localizer[nameof(Resources.Program.Game_ChallengeNotFound)],
+                StatusCodes.Status404NotFound));
+
+        // Check submission limit if configured for this challenge
+        if (context.Challenge!.SubmissionLimit is {} limit)
+        {
+            if (instance.SubmissionCount >= limit)
+                return BadRequest(new RequestResponse(
+                    localizer[nameof(Resources.Program.Challenge_SubmissionLimitExceeded)]));
+        }
+
+        // Increment submission count before creating submission to handle concurrency
+        instance.SubmissionCount++;
```

Comment:
just make a transaction for instance update and submission update, maybe you need concurrence stamp on instance.